CIPs/0005.md

```diff
@@ -0,0 +1,74 @@
+# CIP 0005: Direct Announce
+
+- Date: 2020-02-11
+- Author: @kevjue (with a lot of feedback and ideas from @timmoreton, @m-chrzan, @tkporter)
+- Status: WIP
+
+## Overview
+
+This proposal contains changes to Celo's announce protocol.  That protocol's main goal is to distribute each validator's enodeURL(s) to all the other validators, so that they can create direct TCP connections among each other (effectively create an overlay network among the validators).  This overlay network is then used by Celo's IBFT consensus algorithm to reach consensus on block proposals.
+
+The validator enodeURLs do need to be encrypted when transmitted over the wire, as we want to minimize the number of entities that are aware of it to reduce the change of a DOS attack on the network's validators.  To that end, the enodeURLs needs to be encrypted with each of the remote validators' public key when sent over the wire.  Consequently, each validator will need to share data that is O(N<sup>2</sup>) size where N is the number of elected validators (Note that the validators will also share it's enodeURL(s) with the top 10 validators that are just below the election threshold cutoff.  Throughout this document, the phrase "elected validators" will include those 10).
```

Comment:
> How do we account for a key rotation? We'll want to make sure that if a new validator signer is authorized midway through an epoch, for the rest of the epoch we include the current validator signer & the new validator signer in this set

Good point.  The most straightforward solution is to double the size of the val_enode_table.  Will need to think if there is a more effective solution.

Reply:
> Are connections always maintained amongst the entire set (even those not currently elected)?

Ya, My thinking is since we're reducing the number of nodes that will announce (e.g. it's elected + 10), mind as well make the persisted connection logic easier.

Reply:
> Iirc running an election is pretty expensive of an operation, how often do you expect this set to be calculated? Cached similar to what is done with the active and registered validators?

Yes.  It'll be cached.